Make savetxt() up to 10 times faster for short rows

rudimeier opened the pull request:
In the first 3 commits we are doing a bit cleanup to make the 4th enhancement patch more simple to read. The 5th "refactor structured dtype" is out of my own interest but while I were on it I thought it's just obligatory to add this.

rudimeier committed:
MAINT: savetxt, refactor complex case using vstack

Complex case is now about 50% faster if rows are long but a lot slower
if rows are short. Later we are going to fix the "short rows" case.

rudimeier committed:
MAINT: savetxt, merge complex and real cases

This also fixes 1761a648 where exception handling was
added only for real case.

rudimeier committed:
MAINT: savetxt, muliply strings rather than joining temporary lists

This saves a lot memory if rows are very long.

rudimeier committed:
ENH: savetxt, fast writing short rows

If rows are short we want to format and write many of them at once. See this
benchmarks for N=1000, dtype=float. Speed-up factor 10!

Before:
    N x N:   0.49s
  N*N x 1:   5.12s
N*N/2 x 2:   2.75s
    1 x N*N: 0.53s

After:
    N x N:   0.48s
  N*N x 1:   0.48s
N*N/2 x 2:   0.48s
    1 x N*N: 0.53s

Notes:
* maxel=4000 looks magic but it doesn't seem to matter as long as it's
  between 1000 and 50000. If too large then it needs much more memory
  and it gets slower again.
* We still have another problematic case if rows are very long. This
  patch does not address this one nor does it change the current behavior.
* Now we separate case "complex dtype" more clearly. This case looks
  strange anyways because it works only for 1 dimensional arrays. Could
  be fixed ...

Nodd commented on numpy/lib/npyio.py:
```diff
@@ -1145,21 +1148,37 @@ def savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='',
-                                    "format specifier ('%s')"
-                                    % (str(X.dtype), format))
+
+        # If rows are short we want to format/write up to maxel elemnts at once.
```
Typo, `elemnts` is missing a `e`

rudimeier replied:
On Wednesday 30 September 2015, Joseph Martinot-Lagarde wrote:
> Typo, `elemnts` is missing a `e`

Since adding this `e` overflows 80 chars I've added some more 
comments :)

 # If rows are short we want to format/write up to maxel elements at
 # once (but at least one row, e.g ncol elements).

Nodd replied:
That's a proof that correcting typos is important ! :wink:

rudimeier committed:
MAINT: savetxt, refactor structured dtype

* rename some variables to avoid confusion about complex numbers and
  "complex" (structured) dtype
* BTW structured dtype works also for ndim > 1

Notes:
 * It's still strange that you can't specify a useful fmt. For example for
   dtype=[('x', 'i4'), ('y', 'i4')] you can say fmt="%d" (like it's done
   in our test suite) but for dtype=[('x', 'f8'), ('y', 'i4')] there is no
   way (unless you give a fmt for the whole row). The correct way would be
   to specify fmt="%f %d" but we can't implement this without breaking
   compatibility.

charris commented on numpy/lib/npyio.py:
```diff
@@ -1106,18 +1106,20 @@ def savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='',
-                ncol = len(X.dtype.descr)
-        else:
-            ncol = X.shape[1]
+        nrow = X.shape[0]
```
I don't think this covers all the cases
```
In [28]: dt = dtype([('a','i1'),('b', 'i2'),('','i4')], align=True)

In [29]: a = array([(1,2,3),(4,5,6)], dt)

In [30]: a.shape
Out[30]: (2,)

In [31]: len(a.dtype.descr)
Out[31]: 4

In [32]: b = atleast_2d(a)

In [33]: b.shape
Out[33]: (1, 2)

In [34]: len(b.dtype.descr)
Out[34]: 4
```
The `4` is because an alignment column has been inserted. That may also be a problem, as there are only 3 data columns.

rudimeier replied:
Ok, using len(X.dtype.names), would fix this, right?
--
Sent from myMail app for Android

charris replied:
Not sure. Evidently the tests are incomplete and there haven't been complaints that I am aware of. I am curious as to how alignment columns are handled as they are counted in `len(X.dtype.names)`.

mattip commented:
The PR needs tests. One way to prove it is chunking writes is to create a output stream class with a counter for each write, then assert in the test that the number of writes is as expected, even for recarrays with padding.

charris commented:
I'm going to close this, it is stalled and needs rebase. @rudimeier If you want to pursue this, please make a new PR. It might be worth making the cleanup a separate PR.
